[EuiSuperDatePicker] Add `customQuickSelectRender` prop

breehall opened the pull request:
closes #6378 

## Summary

Created a new `customQuickSelectRender` prop that allows consumers to control the order in which the panels within the Quick Select menu are displayed. This is an optional prop that returns the default panels as args. If the prop is not provided, the panels will appear in the default order (Quick Select, Commonly Used Ranges, Recently Used Ranges, and Custom Panels).

![screencap](https://user-images.githubusercontent.com/549407/208553543-07807920-d07e-432c-9665-be2d8bbced75.gif)

### Usage
`customQuickSelectRender` is used by passing in the Quick Select element in their desired order.

```
customQuickSelectRender={(
      quickSelect,
      commonlyUsedTimes,
      recentlyUsedTimes,
      applyRefreshIntervalPanel,
      customQuickSelectPanels
) => (
   <>
      {customQuickSelectPanels}
      {commonlyUsedTimes}
      {recentlyUsedTimes}
      {quickSelect}
      {applyRefreshIntervalPanel}
   </>
)}
```

### Incidental changes

- Converts `EuiQuickSelectPopover` from a class component to a function component
- Fixes an aria-label missing `i18n` localization, and also makes that label accessible to mouse/sighted users via `title` as well
- Updates quick select popover tests to snapshot rendered components via react-testing-library

## QA

### General checklist

- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
~- [x] Checked for **breaking changes** and labeled appropriately~ No breaking changes
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6382/

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.tsx:
```diff
@@ -92,31 +97,59 @@ export class EuiQuickSelectPopover extends Component<
+        />
+
+        {customQuickSelectRender ? (
+          customQuickSelectRender(
+            commonlyUsedRangesElement,
+            recentlyUsedRangesElement,
+            customQuickSelectPanelsElement
+          )
```
API/naming suggestion - I like being specific with our own internal names, but I think what we pass to the the consumer can be shorter + should be in an obj

```suggestion
          customQuickSelectRender({
            quickSelect: quickSelectElement,
            commonlyUsed: commonlyUsedRangesElement,
            recentlyUsed: recentlyUsedRangesElement,
            customPanels: customQuickSelectPanelsElement,
          })
```

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6382/

constancecchen commented on src/components/date_picker/super_date_picker/super_date_picker.tsx:
```diff
@@ -56,6 +64,13 @@ export type EuiSuperDatePickerProps = CommonProps & {
   customQuickSelectPanels?: QuickSelectPanel[];
 
+  /**
+   * A function that allows the Quick Select panels to render in a custom order
```
Docs feedback - I think we should highlight this specifically as a `render prop` since that's a keyword in React's documentation and is more easily googleable for folks who may be familiar w/ the concept.

I also added in more verbose notes about the flexibility a render prop allows

```suggestion
   * An optional render prop function that allows customizing the display of the Quick Select menu.
   * This function passes all default quick select panels within an object, allowing you to
   * re-order panels, omit certain panels entirely, or pass in your own fully custom content.
```

constancecchen commented on upcoming_changelogs/6382.md:
```diff
@@ -0,0 +1 @@
+- Added the `customQuickSelectRender` prop to `EuiSuperDatePicker` which allows the Quick Select panels to be rendered in a custom order 
```
copy feedback - I tweaked it slightly because this new functionality allows for more than just re-ordering
```suggestion
- Added the `customQuickSelectRender` render prop to `EuiSuperDatePicker`, which allows customizing the Quick Select popover
```

constancecchen commented on src-docs/src/views/super_date_picker/super_date_picker_example.js:
```diff
@@ -218,6 +285,32 @@ if (!endMoment || !endMoment.isValid()) {
+      ],
+      text: (
+        <>
+          <h3>Customing panel order</h3>
```
word nit
```suggestion
          <h3>Customizing panel order</h3>
```

constancecchen replied:
Also, why make this a custom heading inside `text` instead of giving it a `title` and thus allowing it to show up in the sidebar / have an anchor link?

constancecchen commented on src-docs/src/views/super_date_picker/super_date_picker_example.js:
```diff
@@ -218,6 +285,32 @@ if (!endMoment || !endMoment.isValid()) {
+            <EuiCode>customQuickSelectRender</EuiCode> prop that returns Quick
+            Select panels and allows you to customize the order of them.{' '}
+          </p>
+          <EuiBasicTable
+            tableCaption="Quick select panel keys"
+            items={quickSelectTableItems}
+            rowHeader="key"
+            columns={quickSelectTableColumns}
+          />
```
I'm not sure that I love this table or that it's conveying useful information. I'm not sure if this is my personal biases here, but I would personally find a code snippet to be far more useful to actual consumer usage than just listing what we're passing. For example:

```jsx
const CustomQuickSelectPopover = ({
  customQuickSelectPanels,
  quickSelect,
  commonlyUsedTimes,
  recentlyUsedTimes,
  applyRefreshIntervalPanels,
}) => (
  <>
    {customQuickSelectPanels} // Your custom panels, if any
    {commonlyUsedTimes}
    {recentlyUsedTimes}
    {quickSelect} // Normally at the top of the popover
    {applyRefreshIntervalPanels} // Refresh toggles
    // Any of the above passed panels can also be omitted if they aren't wanted,
    // Or you can add your own totally custom JSX/rendering.
  </>
);
```

@miukimiu, do you mind taking a look at the staging docs (https://eui.elastic.co/pr_6382/#/templates/super-date-picker#quick-select-panels) and chiming in here if you can think of a more intuitive way to demo/display this new functionality?

constancecchen replied:
Alternatively... I think the clearest possible way of showing the flexibility of this render prop would be to add some EuiSwitch toggles (1 for each object key), and show/hide the different sections depending on the toggle. I think that's the clearest and most intuitive way to show a 1:1 relationship between the args and what actually appears to users. I'd be happy to pair on this if what I'm asking for is unclear!

Also very optional, but I wonder if it's possible to not including the full EuiSuperDatePicker component for this demo, and if possible, rendering just the quick select popover button/toggle. Honestly, half the time I forget that that button is even there and end up clicking the main date inputs instead.

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.test.tsx:
```diff
@@ -15,8 +15,38 @@ import {
+
 const noop = () => {};
 
+function MyCustomQuickSelectPanel({ applyTime }: { applyTime?: ApplyTime }) {
+  function applyMyCustomTime() {
+    applyTime!({ start: 'now-30d', end: 'now+7d' });
+  }
```
super optional comments/nits:

- The `My` naming feels a little odd to me but I can't quite put my finger on why. I think it's because the distinction between my vs your feels unclear - this could easily be "Your custom panel" instead, for example. Or alternatively, should we just go for a more neutral `CustomQuickSelectPanelExample`? 🤷 I'm probably overtihnking this tbh

- I'd maybe use `() => {}` arrow syntax over `function`. I don't think we're super consistent across EUI but it would be nice to, and arrow fns are generally more modern(?)

constancecchen replied:
Oops, sorry, just saw I left this comment on the wrong file - it was actually intended for [[super_date_picker_custom_quick_select_custom_order.tsx](https://github.com/elastic/eui/pull/6382/files#diff-c648a5c73092615d29d45735e63a30cd0afdddb0e1dbcc952dd1d267c60f09ee)](https://github.com/elastic/eui/pull/6382/files#diff-c648a5c73092615d29d45735e63a30cd0afdddb0e1dbcc952dd1d267c60f09eeR13-R16)

I also would suggest renaming the file to be less about order and more about the new prop being documented specifically, e.g. `super_date_picker_custom_quick_select_render.tsx`

constancecchen reviewed (commented):
Documentation feedback pass - will take a short break and review actual source code changes here in a bit

constancecchen commented on src-docs/src/views/super_date_picker/super_date_picker_custom_quick_select_custom_order.tsx:
```diff
@@ -0,0 +1,99 @@
+            {customQuickSelectPanels}
+            {quickSelect}
+            {commonlyUsedTimes}
+            {recentlyUsedTimes}
```
Leaving here so I don't forget - I haven't examined the code/DOM closely as to why, but recently used ranges/times are not displaying correctly on staging:

<img width="474" alt="" src="https://user-images.githubusercontent.com/549407/204388989-bfda477b-95e5-4672-83e4-ae20714e849c.png">

@breehall just to check, have you pulled in latest main to this branch? We recently had to fix EuiSuperDatePicker margins due to EuiFlex changes (https://github.com/elastic/eui/pull/6380) - not sure if that fix is in this branch

constancecchen replied:
After a bit more digging, I do see that fix is in this branch, so this is a separate issue. It looks like this is caused by default browser styling of `fieldset > legend` components (all browsers apparently silently move/reposition the `legend` if the `legend` is a direct child of `fieldset`.

I see you got around that by adding an `EuiFlexItem` wrapper to the `legend` in commonly used ranges, but I'm not sure that's the best option - I worry that (for some reason) would mess up the semantic relationship/accessibility implications of fieldset/legend.

Can we try using `float: left` ([[1](https://stackoverflow.com/questions/4860708/how-to-position-the-legend-inside-a-fieldset-with-a-border)] [[2](http://sitesbyjoe.com/posts/detail/2011/07/21/styling-fieldsets-and-legends-with-css)]) instead to handle this annoying browser behavior? We'll need to swap the margin-top on `.euiQuickSelectPopover__section` to a margin-bottom on the legend instead:

```scss
.euiQuickSelectPopover__panelTitle {
  float: left;
  margin-bottom: $euiSizeS;
}

// <legend className="euiQuickSelectPopover__paneTitle>
```

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/_quick_select_popover.scss:
```diff
@@ -1,6 +1,16 @@
 .euiQuickSelectPopover__content {
   width: $euiFormMaxWidth;
   max-width: 100%;
+
+  // Add border between customQuickSelect sections
+  > *:first-child {
+    padding-bottom: $euiSizeM;
+  }
+
+  > *:not(:first-child) {
+    border-top: $euiBorderThin;
+    padding: $euiSizeM 0;
+  }
```
Rather than using wildcard child selectors, I would **strongly** suggest that we create our own custom `.euiQuickSelectPopover__panel` class that we apply to the child `fieldset` tag (that currently has no other classes). This way we're also not potentially interfering with any custom rendering consumers are opting to do.

We can also simplify this CSS some by using margins instead of all padding:

```scss
.euiQuickSelectPopover__panel {
  &:not(:first-child) {
    border-top: $euiBorderThin;
    padding-top: $euiSizeM;
    margin-top: $euiSizeM;
  }
}
```

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/commonly_used_time_ranges.tsx:
```diff
@@ -46,14 +46,16 @@ export const EuiCommonlyUsedTimeRanges: FunctionComponent<EuiCommonlyUsedTimeRan
-          />
-        </legend>
-      </EuiTitle>
+      <EuiFlexItem>
+        <EuiTitle size="xxxs">
+          <legend id={legendId}>
+            <EuiI18n
+              token="euiCommonlyUsedTimeRanges.legend"
+              default="Commonly used"
+            />
+          </legend>
+        </EuiTitle>
+      </EuiFlexItem>
```
See https://github.com/elastic/eui/pull/6382#discussion_r1034148863 - I'd prefer we revert this change if possible

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.test.tsx:
```diff
@@ -41,4 +71,105 @@ describe('EuiQuickSelectPopover', () => {
+
+  describe('props', () => {
+    describe('customQuickSelectPanels', () => {
+      it('should render custom panels', () => {
+        const component = shallow(
```
Love these tests, but the snapshot diffs on them were pretty hard to grok. Is there any chance you'd be open to using React Testing Library to `render`, open the popover, and snapshot the DOM contents of the popover instead? No worries if not, this is def an optional request and can be punted to the future

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.test.tsx:
```diff
@@ -41,4 +71,105 @@ describe('EuiQuickSelectPopover', () => {
+          '[data-test-subj="superDatePickerQuickMenu"]'
+        );
+
+        expect(menu.children()).toHaveLength(5);
+
+        expect(menu.children().at(0).is('div')).toBeTruthy();
+
+        expect(menu.children().at(1).is(EuiRefreshInterval)).toBeTruthy();
+
+        expect(menu.children().at(2).is(EuiQuickSelect)).toBeTruthy();
+
+        expect(
+          menu.children().at(3).is(EuiCommonlyUsedTimeRanges)
+        ).toBeTruthy();
+
+        expect(menu.children().at(4).is(EuiRecentlyUsed)).toBeTruthy();
```
Very cool, _love_ that these tests use specific assertions instead of snapshots! 1 very very minor comment is that I'd actually find no newlines between each `expect` easier to read haha. That is my personal preference though so feel free to ignore as an optional suggestion

constancecchen replied:
Also just curious, how useful would you find an optional 3rd unit test that's something like:

```
it('allows omitting passed elements and rendering custom content', () => {
  <EuiQuickSelectPopover customQuickSelectRender={() => <div data-test-subj="customPanel">Hello world</div>} />

  // assert that data-test-subj="customPanel" was rendered
  // assert that the none of the other panels are rendered somehow?
});
```

hmm, maybe this isn't as useful as i thought lol. ah well

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.tsx:
```diff
@@ -84,39 +94,77 @@ export class EuiQuickSelectPopover extends Component<
+        isPaused={isPaused}
+        refreshInterval={refreshInterval}
+      />
+    );
     return (
```
v minor newline nit
```suggestion
    );

    return (
```

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.tsx:
```diff
@@ -84,39 +94,77 @@ export class EuiQuickSelectPopover extends Component<
   };
 
-  renderDateTimeSections = () => {
+  renderQuickSelectMenuSections = () => {
```
Love the rename, and love how clean and easy this is to read! 👏

constancecchen commented on src/components/date_picker/super_date_picker/quick_select_popover/quick_select_popover.tsx:
```diff
@@ -129,26 +177,20 @@ export class EuiQuickSelectPopover extends Component<
       return (
-        <Fragment key={title}>
-          <EuiHorizontalRule margin="s" />
+        <div key={title}>
```
Question - does this need to be a div or can we leave it as a fragment?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6382/

breehall commented:
**Summary of revisions**
- Removed the `EuiFlexItem` wrapper from one of the Quick Select panels and improved the Sass used to create the bottom border and padding between panels. This also resolved a styling bug that caused some of the titles and borders to overlap.
- Updated the `Customizing panel order` section within the docs by removing the table with Quick Select panel keys. Instead, there are now switches that toggle each panel. Hopefully this makes it easier to understand how these panels interact.
- Updated `quick_select_popover` tests to render the entire popover with RTL

I _believe_ that's about everything from the last round of QA!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6382/

constancecchen reviewed (approved):
@breehall I ended up doing a few more major tweaks than minor - apologies! I know you're off for the holidays, but please feel free to ping me after if you have any questions about changes made that I didn't cover in commit messages or was vague about.

Here's the final documentation experience, which I'm hoping you'll agree feels a bit more intuitive/helpful to introduce the render prop concept to consumers:

![screencap](https://user-images.githubusercontent.com/549407/208553543-07807920-d07e-432c-9665-be2d8bbced75.gif)
